designdocs/batched_messages.md

```diff
@@ -10,98 +10,235 @@
-    * `page_size`: `usize`, retrieve `page_size` messages at a time, up to configured `T::BatchPageSizeMax`. If 0, return `T::BatchPageSizeMax` results
-
-* `BatchAnnouncementResult`
-    * `has_next`: `uint32`, current page number
-    * `next_block`: `<T::BlockNumber>` starting block number of next page of results
-    * `next_index`: `u32` starting index of next results in `next_block`
-    * `results`: `Vec<BatchAnnouncement>`
+(See alternatives section for another way to structure payloads)
 
 ### Extrinsics
-#### announce_batch(origin, batch_announcement_params)
-Creates and posts a new batch announcement message on chain, using the batch message Schema Id. This Schema Id must already be registered and will need to be fetched by the extrinsic.  The transaction fee is determined in part by the file size.
+#### register_schema(origin, schema_params)
+Creates and posts a new schema on chain. The transaction fee is determined in part by the model size.
 
 * **Parameters**
   * origin:  required for all extrinsics, the caller/sender.
-  * `batch_announcement_params`: `BatchAnnouncementParams`, the parameters to use in the batch announcement.
+  * `schema_params`: `Schema`, the parameters to use in the batch announcement.
 
-* **Event**:  `Event::<T>::BatchAnnounced(schema_id, msa_id, file_size, batch_uri)`
 * **Restrictions**:
-  * origin's `msa_id` must have capacity to post the transaction during the current epoch
+  * TBD
 
 ### Custom RPCs
 
-#### get_batches(options)
-Retrieves paged batch announcements that have not been garbage-collected which meet `options` criteria.  The `msa_id` does not need to be active.
+#### get_schema(schema_id)
+Retrieves a `Schema`.
 
 * **Parameters**
-  * `options`: `BatchAnnouncementOptions` containing batch announcement criteria
+  * `schema_id`: `u16` a schema identifier
 
 * **Returns**
-  * `None()` if no messages meet the criteria.
-  * `Some(Vec<BatchAnnouncement>)`, in descending block-transaction order
+  * `None()` if no schemas meet the criteria.
+  * `Some(Schema)`
 
+#### MessagesPallet::add(origin, on_behalf_of, schema_id, payload, payload_location)
+This existing RPC call will need to change slightly. The `payload` param, at the
+of this document's writing, is a `Vec<u8>`. This proposal will turn the
+`payload` param's type to `Payload`. It will also add a 5th param for
+`payload_location`.
 
-### Benefits and Risks
-Please see [Batching Source Dependent Messages With Delegation](https://forums.projectliberty.io/t/04-batching-source-dependent-messages-with-delegation/216), for discussion about the benefits of announcing batch files on chain rather than all types of user-created messages.
+* **Parameters**
+  * `origin`: `Origin` A signed transaction origin from the provider
+  * `on_behalf_of`: `Option<MessageSourceId>` The msa id of delegate.
+  * `schema_id`: `u16` A schema identifier
+  * `payload`: `Payload` The message payload
+  * `payload_location`: `PayloadLocation` The payload location
```

Comment:
I'm not sure why we need payload_location. That information should be derivable from `schema_id` right?

Reply:
good point